spanner/client_test.go

```diff
@@ -1480,6 +1480,114 @@ func TestClient_WriteStructWithPointers(t *testing.T) {
 	}
 }
 
+func TestClient_WriteStructWithCustomTypes(t *testing.T) {
+	t.Parallel()
+	server, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+	type CustomString string
+	type CustomBool bool
+	type CustomInt64 int64
+	type CustomFloat64 float64
+	type CustomTime time.Time
+	type CustomDate civil.Date
+	type T struct {
+		ID    int64
+		Col1  CustomString
+		Col2  []CustomString
+		Col3  CustomBool
+		Col4  []CustomBool
+		Col5  CustomInt64
+		Col6  []CustomInt64
+		Col7  CustomFloat64
+		Col8  []CustomFloat64
+		Col9  CustomTime
+		Col10 []CustomTime
+		Col11 CustomDate
+		Col12 []CustomDate
+	}
+	t1 := T{
+		ID:    1,
+		Col2:  []CustomString{},
+		Col4:  []CustomBool{},
+		Col6:  []CustomInt64{},
+		Col8:  []CustomFloat64{},
+		Col10: []CustomTime{},
+		Col12: []CustomDate{},
+	}
+	t2 := T{
+		ID:    2,
+		Col1:  "foo",
+		Col2:  []CustomString{"foo"},
+		Col3:  true,
+		Col4:  []CustomBool{true},
+		Col5:  100,
+		Col6:  []CustomInt64{100},
+		Col7:  3.14,
+		Col8:  []CustomFloat64{3.14},
+		Col9:  CustomTime(time.Now()),
+		Col10: []CustomTime{CustomTime(time.Now())},
+		Col11: CustomDate(civil.DateOf(time.Now())),
+		Col12: []CustomDate{CustomDate(civil.DateOf(time.Now()))},
+	}
+	m1, err := InsertStruct("Tab", &t1)
+	if err != nil {
+		t.Fatal(err)
+	}
+	m2, err := InsertStruct("Tab", &t2)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = client.Apply(context.Background(), []*Mutation{m1, m2})
+	if err != nil {
+		t.Fatal(err)
+	}
+	requests := drainRequestsFromServer(server.TestSpanner)
+	for _, req := range requests {
+		if commit, ok := req.(*sppb.CommitRequest); ok {
+			if g, w := len(commit.Mutations), 2; w != g {
+				t.Fatalf("mutation count mismatch\nGot: %v\nWant: %v", g, w)
+			}
+			insert := commit.Mutations[0].GetInsert()
+			// The first insert should contain empty values and empty arrays
+			for i := 1; i < len(insert.Values[0].Values); i += 2 {
```

Comment:
Can we define a variable for `insert.Values[0]`? Because it has been reused in multiple places.

Reply:
Good point, done.